[CSS Zoom] Fix zoom factor for line-height

vitorroriz opened the pull request:
#### 722ac931d4b8ae3d78668516c604d62d8e215b75
<pre>
[CSS Zoom] Fix zoom factor for line-height
<a href="https://rdar.apple.com/162601748">rdar://162601748</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300699">https://bugs.webkit.org/show_bug.cgi?id=300699</a>

Reviewed by Tim Nguyen and Alan Baradlay.

We want to store the unzoomed line-height  value at RenderStyle
and the zoomeed value at FontCascade.

We are just handling line-height integration of zoom for EvaluationTimeZoomEnabled here.
We are not handling font related units, which is tracked by another bug (<a href="http://webkit.org/b/299816)">http://webkit.org/b/299816)</a>
for that reason, we are explicitly disabling EvaluationTimeZoomEnabled for tests that apply zoom
to elements that present line-height with font-related units (percentage, em, rem...)

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/line-height-basics.html:
* LayoutTests/fast/css/line-height-zoom-get-computed-style.html:
* LayoutTests/fast/text/line-height-minimumFontSize-zoom.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/relative-units.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* LayoutTests/printing/page-rule-selection.html:
* LayoutTests/svg/zoom/page/zoom-replaced-intrinsic-ratio-001.htm:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::preferredLineHeight const):
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
(WebCore::Layout::m_style):
* Source/WebCore/page/PrintContext.cpp:
(WebCore::PrintContext::pageProperty):
* Source/WebCore/rendering/TextAutoSizing.cpp:
(WebCore::TextAutoSizingValue::adjustTextNodeSizes):
* Source/WebCore/rendering/style/AutosizeStatus.cpp:
(WebCore::AutosizeStatus::probablyContainsASmallFixedNumberOfLines):
(WebCore::AutosizeStatus::computeStatus):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::isIdempotentTextAutosizingCandidate const):
(WebCore::RenderStyle::computeLineHeight const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustmentForTextAutosizing):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::zoomWithTextZoomFactor):
* Source/WebCore/style/values/inline/StyleLineHeight.h:
* Source/WebCore/style/values/text/StyleLetterSpacing.cpp:
(WebCore::Style::CSSValueConversion&lt;LetterSpacing&gt;::operator):
* Source/WebCore/style/values/text/StyleWordSpacing.cpp:
(WebCore::Style::CSSValueConversion&lt;WordSpacing&gt;::operator):

Canonical link: <a href="https://commits.webkit.org/301577@main">https://commits.webkit.org/301577@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f03f41c853babdcc813cfd3233d8262be53f838

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126416 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36955 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/133338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/78125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/240452ce-11c0-47a2-adc4-aa931424baf3) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128287 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46719 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54608 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/133338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/78125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129364 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/37368 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/113062 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/133338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/66a8ba8a-bb3c-41db-9dca-f0da33e3774d) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/36258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31240 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76639 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/107176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135869 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53127 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40834 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/104719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53604 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109359 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/104418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26636 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49882 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28213 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50526 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53041 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58862 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52336 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55673 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54065 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/TextAutoSizing.cpp:
```diff
@@ -156,7 +156,7 @@ auto TextAutoSizingValue::adjustTextNodeSizes() -> StillHasNodes
             },
             [&](const Style::LineHeight::Fixed& fixed) {
-                return Style::evaluate<LayoutUnit>(fixed, Style::ZoomNeeded { }).toInt();
+                return Style::evaluate<LayoutUnit>(fixed, Style::ZoomFactor { 1.0f }).toInt();
```
Is 1 correct here?

vitorroriz replied:
I think so as this is used to set the specified line-height

nt1m commented on Source/WebCore/rendering/TextAutoSizing.cpp:
```diff
@@ -168,7 +168,7 @@ auto TextAutoSizingValue::adjustTextNodeSizes() -> StillHasNodes
         // This calculation matches the line-height computed size calculation in StyleBuilderCustom::applyValueLineHeight().
         int lineHeight = specifiedLineHeight * scaleChange;
-        if (auto fixedLineHeight = lineHeightLength.tryFixed(); fixedLineHeight && fixedLineHeight->resolveZoom(Style::ZoomNeeded { }) == lineHeight)
+        if (auto fixedLineHeight = lineHeightLength.tryFixed(); fixedLineHeight && fixedLineHeight->resolveZoom(Style::ZoomFactor { 1.0f }) == lineHeight)
```
Same question here

nt1m commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2591,7 +2591,7 @@ float RenderStyle::computeLineHeight(const Style::LineHeight& lineHeight) const
         },
         [&](const Style::LineHeight::Fixed& fixed) -> float {
-            return Style::evaluate<LayoutUnit>(fixed, Style::ZoomNeeded { }).toFloat();
+            return Style::evaluate<LayoutUnit>(fixed, Style::ZoomFactor { 1.0f }).toFloat();
```
ditto

jesxilin commented on Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
```diff
@@ -166,6 +167,7 @@ class InlineLevelBox {
     struct Style {
         const FontMetrics& primaryFontMetrics;
         const WebCore::Style::LineHeight& lineHeight;
+        // WebCore::Style::ZoomFactor zoomFactor { 1.0f };
```
I think this was left by mistake?

nt1m reviewed (commented):
Can you expand imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html ?

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
```diff
@@ -31,6 +31,10 @@
     "-webkit-text-stroke-width": {
       "value": "2px",
       "otherValues": ["4px"]
+    },
+    "line-height" : {
+      "value": "7px",
+      "otherValues": ["13px"]
     }
```
nit: please preserve alphabetical ordering in this file

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
```diff
@@ -188,7 +189,7 @@ inline void InlineLevelBox::setHasContent()
 }
 
-inline InlineLayoutUnit InlineLevelBox::preferredLineHeight() const
+inline InlineLayoutUnit InlineLevelBox::preferredLineHeight(const RenderStyle& style) const
```
InlineLevelBox retains a set of style values. Please also retain whatever zoom needs (see c'tor).

vitorroriz committed:
[CSS Zoom] Fix zoom factor for line-height
rdar://162601748
https://bugs.webkit.org/show_bug.cgi?id=300699

Reviewed by Tim Nguyen and Alan Baradlay.

We want to store the unzoomed line-height  value at RenderStyle
and the zoomeed value at FontCascade.

We are just handling line-height integration of zoom for EvaluationTimeZoomEnabled here.
We are not handling font related units, which is tracked by another bug (http://webkit.org/b/299816)
for that reason, we are explicitly disabling EvaluationTimeZoomEnabled for tests that apply zoom
to elements that present line-height with font-related units (percentage, em, rem...)

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/line-height-basics.html:
* LayoutTests/fast/css/line-height-zoom-get-computed-style.html:
* LayoutTests/fast/text/line-height-minimumFontSize-zoom.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/relative-units.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* LayoutTests/printing/page-rule-selection.html:
* LayoutTests/svg/zoom/page/zoom-replaced-intrinsic-ratio-001.htm:
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::preferredLineHeight const):
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
(WebCore::Layout::m_style):
* Source/WebCore/page/PrintContext.cpp:
(WebCore::PrintContext::pageProperty):
* Source/WebCore/rendering/TextAutoSizing.cpp:
(WebCore::TextAutoSizingValue::adjustTextNodeSizes):
* Source/WebCore/rendering/style/AutosizeStatus.cpp:
(WebCore::AutosizeStatus::probablyContainsASmallFixedNumberOfLines):
(WebCore::AutosizeStatus::computeStatus):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::isIdempotentTextAutosizingCandidate const):
(WebCore::RenderStyle::computeLineHeight const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustmentForTextAutosizing):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::zoomWithTextZoomFactor):
* Source/WebCore/style/values/inline/StyleLineHeight.h:
* Source/WebCore/style/values/text/StyleLetterSpacing.cpp:
(WebCore::Style::CSSValueConversion<LetterSpacing>::operator):
* Source/WebCore/style/values/text/StyleWordSpacing.cpp:
(WebCore::Style::CSSValueConversion<WordSpacing>::operator):

Canonical link: https://commits.webkit.org/301577@main
